sources/web/datalab/server.ts

```diff
@@ -173,55 +169,111 @@ function handledAuthenticatedRequest(request: http.ServerRequest,
+      logging.getLogger().info('Got auth code');
+      oauth2Client.getToken(query.code, function (err:any, tokens:any) {
+        if (err) {
+          response.writeHead(403);
+          response.end();
+        } else {
+          logging.getLogger().info('Got tokens');
+          oauth2Client.setCredentials(tokens);
+          // Push them to Jupyter and handle request.
+          persistCredentials(tokens);
+          response.statusCode = 302;
+          response.setHeader('Location', query.state);
+          response.end();
+        }
+      });
+    }
     return;
   }
 
-  // /ping allows the deployer to validate existence.
-  if (path.indexOf('/ping') == 0) {
-    // TODO: Remove support for CORS once the existence checks move to the deployment server.
-    response.writeHead(200, {
-      'Content-Type': 'application/json',
-      'Access-Control-Allow-Origin': '*'
+  // Check if EULA has been accepted; if not go to EULA page.
+  if (path.indexOf('/accepted_eula') == 0) {
+    fs.writeFileSync(configFile, '');
+    var i = parsed_url.search.indexOf('referer=');
+    if (i < 0) {
+      logging.getLogger().info('Accepting EULA, but no referer; returning 500');
+      response.writeHead(500);
+    } else {
+      i += 8;
+      var referer = decodeURI(parsed_url.search.substring(i));
+      logging.getLogger().info('Accepting EULA; return to ' + referer);
+      response.writeHead (302, {'Location': referer})
+    }
+    response.end();
+    return;
+  }
+  if (!fs.existsSync(configFile)) {
+    logging.getLogger().info('No Datalab config; redirect to EULA page');
+    fs.readFile('/datalab/web/static/eula.html', function(error, content) {
+      response.writeHead(200);
+      response.end(content);
     });
-
-    // Respond with an object to singal availability and identity.
-    var pingResponse = {
-      name: appSettings.instanceName,
-      id: appSettings.instanceId
-    };
-    response.end(JSON.stringify(pingResponse));
     return;
   }
 
-  // Check if user has access.
-  var userId = userManager.getUserId(request);
-  auth.checkUserAccess(userId, function(e, hasAccess) {
-    if (e) {
-      response.statusCode = 500;
-      response.end("Authentication failure.");
-      return;
-    }
-    if (hasAccess) {
-      handledAuthenticatedRequest(request, response, path);
-    }
-    else {
+  logging.getLogger().info('REMOTE <' + request.connection.remoteAddress + '>');
+  if (process.env.DATALAB_ENV == 'local') {
+    if (oauth2Client) {
+      // We have done auth before. Refresh the token if necessary.
+      // TODO(gram): don't refresh unless token is near expiry.
+      oauth2Client.refreshAccessToken(function (err: any, credentials: any, response_: any) {
+        if (err) {
+          response.writeHead(403);
+          response.end();
+        } else {
+          // Push them to Jupyter and handle request.
+          oauth2Client.setCredentials(credentials);
+          persistCredentials(credentials);
```

Comment:
The TODO needs to be done. :) Regardless of whether refreshAccessToken actually refreshes the token each time, we can't write to disk for each request.